app/src/main/java/com/nervousfish/nervousfish/activities/RhythmCreateActivity.java

```diff
@@ -167,13 +193,165 @@ public void onNewDataReceivedEvent(final NewDataReceivedEvent event) {
             final Contact contact = (Contact) event.getData();
             try {
                 LOGGER.info("Adding contact to database...");
-                this.serviceLocator.getDatabase().addContact(contact);
+                this.database.addContact(contact);
             } catch (IOException | IllegalArgumentException e) {
                 LOGGER.error("Couldn't get contacts from database", e);
             }
 
             //This needs to be outside of the try catch block
-            dataReceived = contact;
+            this.dataReceived = contact;
+        }
+    }
+
+    /**
+     * Denotes the cluster to which a certain tap interval belongs
+     */
+    private enum Cluster {
+        SHORT,
+        LONG
+    }
+
+    /**
+     * Executes K-means++ clustering on the intervals to divide them into a short and long cluster
+     */
+    public static final class KMeansClusterHelper {
+        private List<Long> intervals;
+        private List<Long> clusterCenter1;
+        private List<Long> clusterCenter2;
+
+        private KMeansClusterHelper() {
+            // Prevent instantiation from outside the package
+        }
+
+        private static List<Long> getIntervals(final List<SingleTap> taps) {
+            final List<Long> intervals = new ArrayList<>(taps.size() - 1);
+            for (int i = 0; i < taps.size() - 1; i++) {
+                intervals.add(taps.get(i + 1).getTimestamp().getTime() - taps.get(i).getTimestamp().getTime());
+            }
+            return intervals;
+        }
+
+        /**
+         * Returns the unique key that corresponds to the taps specified.
+         *
+         * @param taps The taps that should be encoded to a key
+         * @return The unique key that corresponds to the taps
+         */
+        int getEncryptionKey(final List<SingleTap> taps) {
+            this.clusterCenter1 = new ArrayList<>(taps.size());
+            this.clusterCenter2 = new ArrayList<>(taps.size());
+            this.intervals = getIntervals(taps);
+            long clusterCenterMean1 = this.makeAndReturnFirstClusterMean();
+            long clusterCenterMean2 = this.makeAndReturnSecondClusterMean();
+
+            while (!this.intervals.isEmpty()) {
+                this.addClosestTimestampToCluster(clusterCenterMean1, clusterCenterMean2);
+                final ImmutablePair<Long, Long> newClusterCenters = this.recalculateClusterCenters();
+                clusterCenterMean1 = newClusterCenters.getLeft();
+                clusterCenterMean2 = newClusterCenters.getRight();
+            }
+
+            final long breakpoint = this.getBreakpoint();
+            this.intervals = getIntervals(taps);
+            return this.generateKey(breakpoint);
+        }
+
+        private long makeAndReturnFirstClusterMean() {
```

Comment:
@TheBonheurs That's right. Officially you don't need to document private stuff with javadoc because private methods and fields are not visible in the documentation that the javadoc generates. In short, you're writing documentation that won't be visible in the documentation.